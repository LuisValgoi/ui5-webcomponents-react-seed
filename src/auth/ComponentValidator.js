import React from 'react';
import { useGet } from '../hooks/useRequest';

import Constants from '../util/Constants';

const ComponentValidator = ({ allowedAuthorities, authorityKey, children }) => {
  const { data, status } = useGet(Constants.REACT_QUERY.KEYS.GET_USER_LOGGED, 'GET_USER_LOGGED', null);

  const getRoute = () => {
    if (status === Constants.REACT_QUERY.CODES.SUCCESS) {
      const hasAccess = data.data.user[authorityKey].some(permission => allowedAuthorities.includes(permission));
      if (hasAccess) {
        return children;
      }
    }

    return <></>;
  };

  return getRoute();
};

export default ComponentValidator;
