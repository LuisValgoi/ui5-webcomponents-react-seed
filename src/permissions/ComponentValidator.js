import React from 'react';
import { useQuery } from 'react-query';

import Request from '../util/Request';
import Constants from '../util/Constants';
import UrlProvider from '../util/URLProvider';

const fetchUser = async () => {
  const url = UrlProvider.getUrl('GET_USER_LOGGED');
  const res = await Request.get(url)
  return res.data;
};

const ComponentValidator = ({ allowedPermissions, allowenceKey, children }) => {
  const { data, status } = useQuery('GET_USER_LOGGED', fetchUser);

  const getRoute = () => {
    if (status === Constants.CODES.RQ_SUCCESS) {
      const hasAccess = data.data.user[allowenceKey].some(permission => allowedPermissions.includes(permission));
      if (hasAccess) {
        return children;
      }
    }

    return <></>;
  };

  return getRoute();
};

export default ComponentValidator;
