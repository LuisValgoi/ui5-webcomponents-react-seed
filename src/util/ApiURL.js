export default {
  GET_USER_LOGGED: '/v1/user/logged',
  GET_TODO_LIST: '/v1/todo/all',
  GET_TODO_BY_ID: '/v1/todo/:id'
};
