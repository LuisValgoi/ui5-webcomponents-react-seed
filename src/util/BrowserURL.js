export default {
  HOME: '/',
  TODO_LIST: '/todo/list',
  TODO_DETAIL: '/todo/:id/detail',
  NOT_FOUND: '/notFound',
  BUGGY: '/buggy',
  ANY: '/*'
};
