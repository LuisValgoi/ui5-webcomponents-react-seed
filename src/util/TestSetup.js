import React from 'react';

import { render as rtlRender } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { rest } from 'msw';
import { setupServer } from 'msw/node';

import BrowserProvider from './browser/BrowserProvider';

const render = (ui, { route = BrowserProvider.getUrl('HOME'), ...renderOptions } = {}) => {
  const WrapperProvider = ({ ...props }) => {
    const history = createMemoryHistory({ initialEntries: [route] });
    return (
      <BrowserRouter>
        <Router history={history}>{props.children}</Router>
      </BrowserRouter>
    );
  };

  return rtlRender(ui, { wrapper: WrapperProvider, ...renderOptions });
};

const serverCustom = (apiUrl, data) => {
  return setupServer(
    rest.get(`*${apiUrl}`, (req, res, ctx) => {
      return res(ctx.json(data));
    }),
  );
};

export * from '@testing-library/react';

export { render, serverCustom };
