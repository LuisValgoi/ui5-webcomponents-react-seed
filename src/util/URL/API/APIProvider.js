import Provider from '../Provider';

const URLs = {
  GET_USER_LOGGED: '/v1/user/logged',
  GET_TODO_BY_ID: '/v1/todo/:id',
  GET_TODO_LIST: '/v1/todo/all',
};

export default {
  getUrl(key, replaceOptions) {
    return Provider.getUrl(URLs, key, replaceOptions);
  },
};
