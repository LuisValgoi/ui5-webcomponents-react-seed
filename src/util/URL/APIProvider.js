const URLs = {
  GET_USER_LOGGED: '/v1/user/logged',
  GET_TODO_BY_ID: '/v1/todo/:id',
  GET_TODO_LIST: '/v1/todo/all',
};

export default {
  getUrl(key) {
    const url = URLs[key];

    if (!url) {
      throw new Error('Url defined by Constant: ' + key + ' not found in APIProvider.js');
    }

    return url;
  },

  replace(url, replaceValue, searchValue = ':id') {
    return url.replace(searchValue, replaceValue);
  },
};
