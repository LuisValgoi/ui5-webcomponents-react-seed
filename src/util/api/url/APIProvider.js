import URLProvider from '../../URLProvider';

const URLs = {
  GET_USER_LOGGED: '/v1/user/logged',
  GET_TODO_BY_ID: '/v1/todo/:id',
  GET_TODO_LIST: '/v1/todo/all',
};

export default {
  getUrl(key, replaceOptions) {
    return URLProvider.getUrl(URLs, key, replaceOptions);
  },
};
