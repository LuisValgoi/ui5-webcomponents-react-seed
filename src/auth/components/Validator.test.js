import React from 'react';

import '@testing-library/jest-dom/extend-expect';
import { render, waitFor, screen, serverCustom } from '../../util/TestSetup';
import ComponentValidator from './Validator';
import APIProvider from '../../util/api/url/APIProvider';

describe('Validator.js Test Suite', () => {
  const GET_USER_LOGGED_RESPONSE = {
    id: 'UG9rZW1vbjowMDE=',
    name: 'testeeeeeeeeeeeeeeeee',
    permissions: ['canAccessTodoListPage', 'canAccessTodoEditPage', 'canAccessDropApplication'],
  };

  const server = serverCustom(APIProvider.getUrl('GET_USER_LOGGED'), GET_USER_LOGGED_RESPONSE);

  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  test('should appear in the document', async () => {
    const childText = 'inner text';

    render(
      <ComponentValidator allowedAuthorities={['canAccessDropApplication']} authorityKey="permissions">
        <p>{childText}</p>
      </ComponentValidator>,
      { route: '/todo/all' },
    );
    const child = await waitFor(() => screen.getByText(childText));

    expect(child).toBeInTheDocument();
  });

  test.only('should not appear in the document', async () => {
    const childText = 'inner text';

    render(
      <ComponentValidator allowedAuthorities={['canAccessDropApplication']} authorityKey="permissions">
        <p>{childText}</p>
      </ComponentValidator>,
      { route: '/todo/all' },
    );
    console.log('PATHHHHHH', window.location.pathname);
    const child = screen.queryByAltText(childText);

    expect(child).not.toBeInTheDocument();
  });
});
