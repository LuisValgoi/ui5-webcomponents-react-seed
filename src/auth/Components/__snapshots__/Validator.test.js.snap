// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validator.js Test Suite should match snapshot 1`] = `
<p>
  inner text
</p>
`;
